sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/ReadFromKafkaDoFn.java

```diff
@@ -288,6 +316,19 @@ public ProcessContinuation processElement(
               Optional.ofNullable(watermarkEstimator.currentWatermark()));
     }
     try (Consumer<byte[], byte[]> consumer = consumerFactoryFn.apply(updatedConsumerConfig)) {
+      // Check whether current TopicPartition is still available to read.
+      Set<TopicPartition> existingTopicPartitions = new HashSet<>();
+      for (List<PartitionInfo> topicPartitionList : consumer.listTopics().values()) {
```

Comment:
How often `consumer.listTopics()` will be called? Won't it affect a performance? Will it make sense to cache the result?

Reply:
For one bundle, it is called once per `KafkaSourceDescriptor`. It has been a result compromising for performance concern.

If we want to get accurate behavior(stop emitting records as soon as the TopicPartition is stopped/removed), we should call `listTopics()` per record. But we know that it will affect performance a lot.

We can also do caching but it will reduce more accuracy.

Reply:
Ok, let's keep it as it is for now.